File: python/src/arkadia/ai/data/Encoder.py
# encoder.py
"""
Encoder for AI.DATA-FORMAT.

Supports:
- Primitive values
- Record {key:value}
- Lists [ ... ]
- Lists of primitives
- Inline type headers <a:number b:string c:string[]>
"""

from email import header
from typing import Any, Optional, Union
from xml.etree.ElementTree import indent
from .Node import Node
from .Schema import Schema, SchemaKind
from .Meta import Meta
from .Config import Config


class Colors:
    RESET = "\033[0m"
    STRING = "\033[92m"  # green
    NUMBER = "\033[94m"  # blue
    BOOL = "\033[95m"  # magenta
    NULL = "\033[90m"  # gray
    TYPE = "\033[96m"  # cyan
    KEY = "\033[93m"  # yellow
    SCHEMA = "\033[91m"  # red (for @TypeName)
    TAG = "\033[91m"
    ATTR = "\033[93m"



class Encoder:
    def __init__(self, config: Config):
        self.indent = config.get("indent", 2)
        self.start_indent = config.get("start_indent", 0)
        self.compact = config.get("compact", False)
        self.escape_new_lines = config.get("escape_new_lines", False)
        self.colorize = config.get("colorize", False)
        self.include_comments = config.get("include_comments", True)
        self.include_array_size = config.get("include_array_size", False)
        self.include_schema = config.get("include_schema", True)
        self.include_type = config.get("include_type", True)
        self.include_meta = config.get("include_meta", True)
        # This is special encode, when the schema is icnluded inside sample data
        # [
        # (name: string /name of the user/, age: number /age of the user/),
        # ]
        self.prompt_output = config.get("prompt_output", False)

    # -------------------------------------------------------------
    # PUBLIC ENTRY
    # -------------------------------------------------------------

    def encode(self, node: Node, indent: int = 0, include_schema: bool = True) -> str:
        """
        Encode a Node into ADF text.
        Order: Meta/Comments -> Schema -> Data
        """
        base_indent = self.start_indent + indent

        # 2. Prepare Schema Header (e.g. <test: string>)
        schema_prefix = ""
        should_render_schema = (
            include_schema 
            and node.schema is not None 
            and self.include_schema # Config check
        )
        
        if should_render_schema:
            s_txt = self.encode_schema(node.schema, base_indent).strip()

            if  not s_txt.startswith("<") and not s_txt.startswith("@"): 
                s_txt = f"<{s_txt}>"

            if s_txt:
                if self.compact:
                    schema_prefix = s_txt
                else:
                    # Newline formatting for schema header
                    schema_prefix = s_txt + "\n" + (" " * base_indent)
                    
        # 3. Encode Data
        if node.is_list:
            data = self._list(node, base_indent, include_schema=False)

        elif node.is_primitive:
            data = self._primitive_node(node)
        
        elif node.is_record:
            data = self._record(node, base_indent)
        
        else:
            data = self._c("null", Colors.NULL)

        # Final Assembly: Meta -> Schema -> Data
        return f"{schema_prefix}{data}" 



    def encode_schema(self, schema: Schema, indent: int = 2, include_meta: bool = True) -> str:
        """
        Encode Schema into AI.Data header form.
        Supports: < [ ... ] >, / meta /, !required
        """
        if schema is None:
            return ""

        ind = " " * indent
        prefix = ""
        
        # Determine padding for brackets based on compact mode
        pad = "" if self.compact else " "

        # Avoid printing internal/default type names
        if (
            schema.type_name
            and schema.kind == SchemaKind.RECORD
            and not schema.is_any
        ):
            prefix = self._c(f"@{schema.type_name}", Colors.SCHEMA)

        meta = self._meta_wrapped(schema) if include_meta else ""

        # --- PRIMITIVE ---
        if schema.is_primitive:
            meta_prefix = self._meta_inline(schema) if include_meta else ""
            return ind + ((meta_prefix + " ") if meta_prefix else "") + self._c(schema.type_name, Colors.TYPE)

        # --- LIST ---
        if schema.is_list:
            # Check for Outer Meta for the list itself
            schema.apply_meta(schema.element)
            schema.element.clear_meta()
            list_meta = self._meta_wrapped(schema) if include_meta else ""
            
            # Special Case: List of Records < [ ... ] >
            if schema.element and schema.element.is_record:
                # We reuse the _record_fields logic but wrap in <[ ... ]>
                inner_fields = self._encode_schema_fields(schema.element)
    
                # FIX: Use 'pad' variable to remove spaces in compact mode
                return ind + prefix + "<" + pad + "[" + list_meta + inner_fields + pad + "]" + pad + ">"

            # Standard List [Type]
            inner = self.encode_schema(schema.element, 0, False).strip()
            return ind + "[" + list_meta + self._c(inner, Colors.TYPE) + "]"

        # --- RECORD ---
        if schema.is_record:
            # Get Record-level meta (e.g. < / $ver=1 / ... >)
            record_meta = self._meta_wrapped(schema) if include_meta else ""

            if not schema.fields:
                # If the record is generic (no fields, no specific type name, no meta),
                # return an empty string to avoid printing "<>" or "<any>" before the "{...}".
                if not prefix and not record_meta and schema.is_any:
                    return ""
                # FIX: Use 'pad' variable
                return ind + prefix + "<" + pad + record_meta + "any" + pad + ">"

            # Encode Fields
            inner_fields = self._encode_schema_fields(schema)

            # FIX: Use 'pad' variable to remove spaces in compact mode
            # e.g., Compact: <id:int> | Pretty: < id: int >
            return ind + prefix + "<" + pad + record_meta + inner_fields + pad + ">"

        return ind + f"<{meta if include_meta else ''}any>"

    # --- Helper to deduplicate field encoding logic ---
    def _encode_schema_fields(self, schema: Schema) -> str:
        """
        Encodes fields definitions.
        Format: /* comment */ !required $attr=val #tag name: type
        """
        parts = []
        pad = "" if self.compact else " "

        for field in schema.fields:
            field_parts = []

            meta_prefix = self._meta_inline(field)
            if meta_prefix:
                field_parts.append(meta_prefix)

            # 3. Field Name
            field_parts.append(self._c(field.name, Colors.KEY))

            # 4. Field Type
            field_type = self.encode_schema(field, 0, False).strip()
            
            last_idx = len(field_parts) - 1
            
            # FIX: Logic to decide when to print the type signature.
            # A. If it's a Structure (List/Record), ALWAYS print the type (e.g. [string]), 
            #    ignoring the fact that type_name might be default "any".
            # B. If it's a Primitive, print it only if it's not "any" (and types are included).
            # C. Crucial: Ensure field_type is not empty (handles generic empty records).
            is_structure = not field.is_primitive
            is_explicit_primitive = (self.include_type and field.type_name != "any")
            
            if field_type and (is_structure or is_explicit_primitive):
                field_parts[last_idx] += f":{self._c(field_type, Colors.TYPE)}"
            
            parts.append(" ".join(field_parts))

        sep = f",{pad}"
        return sep.join(parts)


    # -------------------------------------------------------------
    # HELPER: SCHEMA COMPATIBILITY CHECK
    # -------------------------------------------------------------
    
    def _schemas_are_compatible(self, node_schema: Optional[Schema], expected_schema: Optional[Schema]) -> bool:
        """
        Checks if the node's schema matches the parent's expected schema.
        Used to determine if we need to show an inline <type> override.
        """
        if not expected_schema or expected_schema.is_any:
            return True
        
        if node_schema is None:
            return True
        
        # Check general kind
        if node_schema.kind != expected_schema.kind:
            return False
            
        # Check specific primitive type name (e.g. int vs string)
        if node_schema.is_primitive and expected_schema.is_primitive:
            # Loose equality: 'number' matches 'int'/'float' usually, but here we check exact name match
            return node_schema.type_name == expected_schema.type_name
            
        return True

    def _get_type_label(self, schema: Schema) -> str:
        """Generates a short type label for inline overrides <type>."""
        if schema.is_primitive:
            return schema.type_name
        elif schema.is_list:
            inner = self._get_type_label(schema.element)
            return f"[{inner}]"
        elif schema.is_record and schema.type_name and schema.type_name != "any":
            return schema.type_name
        # Fallback for complex inline records
        return "any"

    def _apply_type_tag(self, val_str: str, node_schema: Optional[Schema], expected_schema: Optional[Schema]) -> str:
        """
        Applies an inline type tag (e.g., <int>) if the schema does not match the expectation.
        This is the single source of truth for override formatting.
        """
        if self._schemas_are_compatible(node_schema, expected_schema):
            return val_str
        
        # Mismatch detected -> Wrap with tag
        label = self._get_type_label(node_schema) if node_schema else "any"
        tag = self._c(f"<{label}>", Colors.SCHEMA)
        return f"{tag} {val_str}"


    # -------------------------------------------------------------
    # META AND COMMENTS (Unified Logic)
    # -------------------------------------------------------------

    def _build_meta_string(self, obj: Meta) -> str:
        """
        Internal helper to build the content of metadata string.
        Combines Comments, Attributes, Tags, and Modifiers.
        """
        items = []
        pad =  ""  if self.compact else " "

        # 1. Comments
        if self.include_comments and obj.comments:
            for c in obj.comments:
                cleaned = c.strip()
                items.append(self._c(f"/*{pad}{cleaned}{pad}*/", Colors.NULL))

        # 2. Modifiers
        if getattr(obj, 'required', False):
            items.append(self._c("!required", Colors.TAG))

        # 3. Attributes & Tags
        if self.include_meta:
            current_attr = obj.attr or {}
            for k, v in current_attr.items():
                val_str = self._primitive(v)
                items.append(self._c(f"${k}=", Colors.ATTR) + val_str)

            current_tags = obj.tags or []
            for t in current_tags:
                items.append(self._c(f"#{t}", Colors.TAG))

        if not items:
            return ""

        content = " ".join(items)
        
        # Inline: /* c1 */ !req $a=1
        return content

    def _meta_inline(self, obj: Meta) -> str:
        """For Primitives/Fields: No wrappers."""
        return self._build_meta_string(obj)
    
    def _meta_wrapped(self, obj: Meta) -> str:
        """For Containers/Schema Headers: Wrapped in /.../."""
        content = self._build_meta_string(obj)
        return self._wrap_meta(content)

    def _wrap_meta(self, content: str):
        if not content:
            return ""
        pad =  ""  if self.compact else " "
        content = self._c(f"/{pad}", Colors.SCHEMA) + content + self._c(f"{pad}/", Colors.SCHEMA)
        if self.compact:
            return content + " " if content else ""
        else:
            return " " + content + " " if content else ""
   
    # -------------------------------------------------------------
    # HELPERS
    # -------------------------------------------------------------
    def _c(self, text: str, color: str) -> str:
        if not self.colorize:
            return text
        return f"{color}{text}{Colors.RESET}"

    def _escape_newlines(self, text: str) -> str:
        return text.replace("\r\n", "\\n").replace("\r", "\\n").replace("\n", "\\n")

    def _primitive_node(self, node: Node) -> str:
        inner_meta = self._meta_inline(node)
        str = self._primitive(node.value)
        return (inner_meta + " " if inner_meta else "") + str
        

    def _primitive(self, v: Any) -> str:
        if isinstance(v, str):
            return self._string(v)
        if v is True:
            return self._c("true", Colors.BOOL)
        if v is False:
            return self._c("false", Colors.BOOL)
        if v is None:
            return self._c("null", Colors.NULL)
        return self._c(str(v), Colors.NUMBER)

    def _string(self, v: str) -> str:
        content = v
        if self.escape_new_lines:
            content = self._escape_newlines(content)
        content = content.replace('"', '\\"')
        return self._c(f'"{content}"', Colors.STRING)

    # -------------------------------------------------------------
    # PRIMITIVE LIST: ["a","b","c"]
    # -------------------------------------------------------------

    def _list_header(self, node: Node) -> str:
        header = "["
        if self.include_array_size:
            # Prefer elemeents length if available
            size = (
                len(node.elements)
            )
            header += f"{self._c('$size', Colors.KEY)}={self._c(str(size), Colors.NUMBER)}{self._c(':', Colors.TYPE)}"
        return header

    # -------------------------------------------------------------
    # STRUCTURAL LIST: [ Node, Node ]
    # -------------------------------------------------------------
    def _join(self, items: list[str], sep: str) -> str:
        if self.compact:
            return sep.join(items)

        if sep == "\n":
            return sep.join(items)

        # For inline separators (like commas), we add a space for readability
        return f"{sep} ".join(items)

    # -------------------------------------------------------------
    # STRUCTURAL LIST: [ @Schema... (...) (...) ]
    # -------------------------------------------------------------
    def _list(self, node: Node, indent: int, include_schema: bool = False) -> str:
        ind = " " * indent
        child_indent = indent + self.indent

        inner_meta = self._meta_wrapped(node)

        # 1. Generate Header Schema (if requested)
        schema_header = ""
        if include_schema and node.schema is not None and node.schema.element is not None:
            # This generates the <[ id: int ]> part
            schema_header = self.encode_schema(node.schema.element, 0).strip()
        if schema_header:
            schema_header = schema_header + " "

        expected_child = node.schema.element if node.schema else None

        # --- COMPACT MODE ---
        if self.compact:
            items = []

            
            for el in node.elements:
                # IMPORTANT: We disable schema inclusion for elements to avoid duplication <...>
                # unless types mismatch drastically.
                val = self.encode(el, 0, include_schema=False).strip()

                # Check compatibility & Inject override if needed
                if not self._schemas_are_compatible(el.schema, expected_child):
                    label = self._get_type_label(el.schema)
                    tag = self._c(f"<{label}>", Colors.SCHEMA)
                    val = f"{tag} {val}"

                items.append(val)

            return ind + "[" + inner_meta + schema_header  + ",".join(items) + "]"

        # --- PRETTY MODE ---
        header = self._list_header(node)
        out = [ind + header]

        if inner_meta:
            out.append(" " * child_indent + inner_meta)

        if schema_header:
            out.append(" " * child_indent + schema_header)

        for el in node.elements:
            # IMPORTANT: Disable schema for children
            val = self.encode(
                el, child_indent - self.start_indent, include_schema=False
            )
            val = val.strip()

            # Check compatibility & Inject override if needed
            if not self._schemas_are_compatible(el.schema, expected_child):
                label = self._get_type_label(el.schema)
                tag = self._c(f"<{label}>", Colors.SCHEMA)
                val = f"{tag} {val}"

            out.append(" " * child_indent + val)

        out.append(ind + "]")
        return self._join(out, "\n")
    
  
    # -------------------------------------------------------------
    # RECORD: {key:value} -> (val1, val2) OR {key:value}
    # -------------------------------------------------------------
    def _record(self, node: Node, indent: int) -> str:
        inner_meta = self._meta_wrapped(node)

        parts = []
        if node.schema.fields:
            for field_def in node.schema.fields:
                field_node = node.fields.get(field_def.name)
                if field_node:
                    val = self.encode(
                        field_node, indent - self.start_indent, False
                    ).strip()
                    val = self._apply_type_tag(val, field_node.schema, field_def)
                    parts.append(val)
                else:
                    parts.append(self._c("null", Colors.NULL))
        else:
            parts.append(self._c("null", Colors.NULL))

        sep = ", " if not self.compact else ","
        return "(" + inner_meta + sep.join(parts) + ")"

File: typescript/src/core/Encoder.ts
import { EncoderConfig } from '../config';
import { Meta } from '../models/Meta';
import { Node, Primitive } from '../models/Node';
import { Schema } from '../models/Schema';

class Colors {
  static RESET = '\x1b[0m';
  static STRING = '\x1b[92m'; // green
  static NUMBER = '\x1b[94m'; // blue
  static BOOL = '\x1b[95m'; // magenta
  static NULL = '\x1b[90m'; // gray
  static TYPE = '\x1b[96m'; // cyan
  static KEY = '\x1b[93m'; // yellow
  static SCHEMA = '\x1b[91m'; // red (for @TypeName)
  static TAG = '\x1b[35m';
  static ATTR = '\x1b[93m';
}

export class Encoder {
  private config: EncoderConfig;

  constructor(config: EncoderConfig) {
    this.config = config;
  }

  // -------------------------------------------------------------
  // PUBLIC ENTRY
  // -------------------------------------------------------------

  public encode(node: Node, indentLevel: number = 0, includeSchema: boolean = true): string {
    const baseIndent = this.config.startIndent + indentLevel;
    let schemaPrefix = '';

    // 2. Prepare Schema Header (e.g. <test: string>)
    const shouldRenderSchema = includeSchema && node.schema !== null && this.config.includeSchema;

    if (shouldRenderSchema) {
      let sTxt = this.encodeSchema(node.schema!, baseIndent).trim();

      // Jeśli wynik nie jest pusty i nie zaczyna się od < ani @, dodaj nawiasy < >
      if (sTxt && !sTxt.startsWith('<') && !sTxt.startsWith('@')) {
        sTxt = `<${sTxt}>`;
      }

      if (sTxt) {
        if (this.config.compact) {
          schemaPrefix = sTxt;
        } else {
          // Newline formatting for schema header
          schemaPrefix = sTxt + '\n' + ' '.repeat(baseIndent);
        }
      }
    }

    // 3. Encode Data
    let data = '';
    if (node.isList) {
      data = this.encodeList(node, baseIndent, false);
    } else if (node.isPrimitive) {
      data = this.encodePrimitiveNode(node);
    } else if (node.isRecord) {
      data = this.encodeRecord(node, baseIndent);
    } else {
      data = this.c('null', Colors.NULL);
    }

    // Final Assembly: Meta -> Schema -> Data
    return `${schemaPrefix}${data}`;
  }

  public encodeSchema(schema: Schema, indent: number, includeMeta: boolean = true): string {
    if (!schema) return '';

    const ind = ' '.repeat(indent);
    let prefix = '';

    const pad = this.config.compact ? '' : ' ';

    // Avoid printing internal/default type names
    if (schema.typeName && schema.isRecord && !schema.isAny) {
      const escapedName = this.escapeIdent(schema.typeName);
      prefix = this.c(`@${escapedName}`, Colors.SCHEMA);
    }

    // Przygotowanie meta (ale jeszcze nie użycie, bo w liście może się zmienić)
    // W Pythonie: meta = self._meta_wrapped(schema) if include_meta else ""
    // Tutaj obliczamy dynamicznie tam gdzie potrzeba, bo w liście modyfikujemy obiekt schema.

    // --- PRIMITIVE ---
    if (schema.isPrimitive) {
      const metaPrefix = includeMeta ? this.metaInline(schema) : '';
      // Python: return ind + ((meta_prefix + " ") if meta_prefix else "") + self._c(...)
      const metaStr = metaPrefix ? metaPrefix + ' ' : '';
      const escapedType = this.escapeIdent(schema.typeName);
      return ind + metaStr + this.c(escapedType, Colors.TYPE);
    }

    // --- LIST ---
    // --- DEBUG ---
    if (schema.isList) {
      if (schema.element) {
        schema.applyMeta(schema.element);
        schema.element.clearMeta();
      }
      const listMeta = includeMeta ? this.metaWrapped(schema) : '';

      // Special Case: List of Records < [ ... ] >
      if (schema.element && schema.element.isRecord) {
        // We reuse the _record_fields logic but wrap in <[ ... ]>
        const innerFields = this.encodeSchemaFields(schema.element);

        // FIX: Use 'pad' variable to remove spaces in compact mode
        return ind + prefix + '<' + pad + '[' + listMeta + innerFields + pad + ']' + pad + '>';
      }

      // Standard List [Type]
      const inner = this.encodeSchema(schema.element!, 0, false).trim();
      return ind + '[' + listMeta + this.c(inner, Colors.TYPE) + ']';
    }

    // --- RECORD ---
    if (schema.isRecord) {
      // Get Record-level meta (e.g. < / $ver=1 / ... >)
      const recordMeta = includeMeta ? this.metaWrapped(schema) : '';

      if (!schema.fields || schema.fields.length === 0) {
        // If the record is generic (no fields, no specific type name, no meta),
        // return an empty string to avoid printing "<>" or "<any>" before the "{...}".
        if (!prefix && !recordMeta && schema.isAny) {
          return '';
        }
        return ind + prefix + '<' + pad + recordMeta + 'any' + pad + '>';
      }

      // Encode Fields
      const innerFields = this.encodeSchemaFields(schema);

      // FIX: Use 'pad' variable to remove spaces in compact mode
      return ind + prefix + '<' + pad + recordMeta + innerFields + pad + '>';
    }

    // Fallback (ANY)
    const meta = includeMeta ? this.metaWrapped(schema) : '';
    return ind + `<${meta}any>`;
  }

  // --- Helper to deduplicate field encoding logic ---
  private encodeSchemaFields(schema: Schema): string {
    const parts: string[] = [];
    const pad = this.config.compact ? '' : ' ';

    for (const field of schema.fields) {
      let str = '';

      const metaPrefix = this.metaInline(field);
      if (metaPrefix) {
        str += metaPrefix + ' ';
      }

      // 3. Field Name
      str += this.c(this.escapeIdent(field.name), Colors.KEY);

      // 4. Field Type
      const fieldType = this.encodeSchema(field, 0, false).trim();

      // Logic to decide when to print the type signature
      const isStructure = !field.isPrimitive;
      const isExplicitPrimitive = this.config.includeType && field.typeName !== 'any';

      if (fieldType && (isStructure || isExplicitPrimitive)) {
        str += `:${pad}${this.c(fieldType, Colors.TYPE)}`;
      }

      parts.push(str);
    }

    const sep = `,${pad}`;
    return parts.join(sep);
  }

  // -------------------------------------------------------------
  // HELPER: SCHEMA COMPATIBILITY CHECK
  // -------------------------------------------------------------

  private schemasAreCompatible(nodeSchema: Schema | null, expectedSchema: Schema | null): boolean {
    if (!expectedSchema || expectedSchema.isAny) return true;
    if (!nodeSchema) return true;

    // Check general kind
    if (nodeSchema.kind !== expectedSchema.kind) return false;

    // Check specific primitive type name (e.g. int vs string)
    if (nodeSchema.isPrimitive && expectedSchema.isPrimitive) {
      return nodeSchema.typeName === expectedSchema.typeName;
    }

    return true;
  }

  private getTypeLabel(schema: Schema): string {
    if (schema.isPrimitive) return this.escapeIdent(schema.typeName || 'any');
    if (schema.isList) {
      const inner = schema.element ? this.getTypeLabel(schema.element) : 'any';
      return `[${inner}]`;
    }
    if (schema.isRecord && schema.typeName && schema.typeName !== 'any') {
      return this.escapeIdent(schema.typeName);
    }
    return 'any';
  }

  private applyTypeTag(
    valStr: string,
    nodeSchema: Schema | null,
    expectedSchema: Schema | null,
  ): string {
    if (this.schemasAreCompatible(nodeSchema, expectedSchema)) {
      return valStr;
    }

    // Mismatch detected -> Wrap with tag
    const label = nodeSchema ? this.getTypeLabel(nodeSchema) : 'any';
    const tag = this.c(`<${label}>`, Colors.SCHEMA);
    return `${tag} ${valStr}`;
  }

  // -------------------------------------------------------------
  // META AND COMMENTS (Unified Logic)
  // -------------------------------------------------------------

  private buildMetaString(obj: Meta, wrapped: boolean = false): string {
    const items: string[] = [];
    const pad = this.config.compact ? '' : ' ';

    // 1. Comments
    if (this.config.includeComments && obj.comments) {
      for (const c of obj.comments) {
        const cleaned = c.trim();
        items.push(this.c(`/*${pad}${cleaned}${pad}*/`, Colors.NULL));
      }
    }

    // 2. Modifiers
    if ((obj as Schema).required) {
      items.push(this.c('$required', Colors.TAG));
    }

    // 3. Attributes & Tags
    if (this.config.includeMeta) {
      const currentAttr = obj.attr || {};
      for (const [k, v] of Object.entries(currentAttr)) {
        const escapedKey = this.escapeIdent(k);
        if (typeof v === 'boolean' && v === true) {
          items.push(this.c(`$${escapedKey}`, Colors.ATTR));
        } else {
          const valStr = this.primitiveValue(v);
          items.push(this.c(`$${escapedKey}=`, Colors.ATTR) + valStr);
        }
      }

      const currentTags = obj.tags || [];
      for (const t of currentTags) {
        items.push(this.c(`#${t}`, Colors.TAG));
      }
    }

    if (items.length === 0) {
      return '';
    }

    const content = items.join(' ');

    if (wrapped) {
      const wrappedContent =
        this.c(`//${pad}`, Colors.SCHEMA) + content + this.c(`${pad}//`, Colors.SCHEMA);
      return this.config.compact ? wrappedContent + ' ' : ' ' + wrappedContent + ' ';
    }

    // Inline: /* c1 */ !req $a=1
    return content;
  }

  private metaInline(obj: Meta): string {
    return this.buildMetaString(obj, false);
  }

  private metaWrapped(obj: Meta): string {
    return this.buildMetaString(obj, true);
  }

  // -------------------------------------------------------------
  // HELPERS
  // -------------------------------------------------------------
  private c(text: string, color: string): string {
    if (!this.config.colorize) return text;
    return `${color}${text}${Colors.RESET}`;
  }

  private encodePrimitiveNode(node: Node): string {
    const innerMeta = this.metaInline(node);
    const str = this.primitiveValue(node.value as Primitive);
    return (innerMeta ? innerMeta + ' ' : '') + str;
  }

  private primitiveValue(v: Primitive): string {
    if (typeof v === 'string') return this.encodeString(v);
    if (v === true) return this.c('true', Colors.BOOL);
    if (v === false) return this.c('false', Colors.BOOL);
    if (v === null || v === undefined) return this.c('null', Colors.NULL);
    return this.c(String(v), Colors.NUMBER);
  }

  private encodeString(v: string): string {
    // 1. Escape backslashes first! (Critical step)
    // We must turn a single literal \ into \\ before processing quotes.
    let content = v.replace(/\\/g, '\\\\');

    // 2. Escape quotes
    // Turn " into \"
    content = content.replace(/"/g, '\\"');

    // 3. Handle newlines and control characters if configured
    if (this.config.escapeNewLines) {
      content = content.replace(/\n/g, '\\n').replace(/\r/g, '\\r').replace(/\t/g, '\\t');
    }

    // Wrap in quotes and colorize
    return this.c(`"${content}"`, Colors.STRING);
  }

  /**
   * Wraps identifier in backticks if it contains spaces or special characters.
   * Regex: [a-zA-Z_][a-zA-Z0-9_]*
   */
  private escapeIdent(name: string): string {
    if (!name) return '';

    // Standard identifier pattern
    const pattern = /^[a-zA-Z_][a-zA-Z0-9_]*$/;

    if (pattern.test(name)) {
      return name;
    }

    // If it doesn't match, wrap it in backticks
    return `\`${name}\``;
  }

  // -------------------------------------------------------------
  // LIST
  // -------------------------------------------------------------

  // private listHeader(node: Node): string {
  //   let header = '[';
  //   if (this.config.includeArraySize) {
  //     const size = node.elements.length;
  //     header += `${this.c('$size', Colors.KEY)}=${this.c(String(size), Colors.NUMBER)}${this.c(':', Colors.TYPE)}`;
  //   }
  //   return header;
  // }

  // private joinLines(items: string[], sep: string): string {
  //   if (this.config.compact) return items.join(sep);
  //   if (sep === '\n') return items.join(sep);
  //   return items.join(`${sep} `);
  // }

  private encodeList(node: Node, indent: number, includeSchema: boolean = false): string {
    const ind = ' '.repeat(indent);
    const childIndent = indent + this.config.indent;

    if (this.config.includeArraySize) {
      node.attr['size'] = node.elements.length;
    }
    const innerMeta = this.metaWrapped(node);

    // 1. Generate Header Schema (if requested)
    let schemaHeader = '';
    if (includeSchema && node.schema && node.schema.element) {
      schemaHeader = this.encodeSchema(node.schema.element, 0).trim();
    }
    if (schemaHeader) {
      schemaHeader = schemaHeader + ' ';
    }

    const expectedChild = node.schema ? node.schema.element : null;

    // --- COMPACT MODE ---
    if (this.config.compact) {
      const items: string[] = [];

      for (const el of node.elements) {
        // IMPORTANT: We disable schema inclusion for elements to avoid duplication <...>
        // unless types mismatch drastically.
        let val = this.encode(el, 0, false).trim();

        // Check compatibility & Inject override if needed
        if (!this.schemasAreCompatible(el.schema, expectedChild)) {
          const label = el.schema ? this.getTypeLabel(el.schema) : 'any';
          const tag = this.c(`<${label}>`, Colors.SCHEMA);
          val = `${tag} ${val}`;
        }

        items.push(val);
      }

      return ind + '[' + innerMeta + schemaHeader + items.join(',') + ']';
    }

    // --- PRETTY MODE ---
    let res = ind + '[\n';

    // 4. Add Meta-data line (isolated from commas to avoid syntax errors)
    if (innerMeta) {
      res += ' '.repeat(childIndent) + innerMeta.trim() + '\n'; // cite: 1.1
    }

    // 5. Add Schema Header line (isolated from commas)
    if (schemaHeader) {
      res += ' '.repeat(childIndent) + schemaHeader.trim() + '\n'; // cite: 2.2
    }

    // 6. Generate and collect element lines
    const elementLines: string[] = [];
    for (const el of node.elements) {
      // We encode with 0 indent because we prepend the childIndent manually
      let val = this.encode(el, 0, false).trim();

      // Check for polymorphic types and add schema tags where necessary
      if (!this.schemasAreCompatible(el.schema, expectedChild)) {
        const label = el.schema ? this.getTypeLabel(el.schema) : 'any';
        const tag = this.c(`<${label}>`, Colors.SCHEMA);
        val = `${tag} ${val}`;
      }

      elementLines.push(' '.repeat(childIndent) + val);
    }

    // 7. Join elements with commas and newlines only (preserving metadata integrity)
    if (elementLines.length > 0) {
      res += elementLines.join(',\n') + '\n'; // cite: 1.1
    }

    // 8. Close the list block
    res += ind + ']';
    return res;
  }

  // -------------------------------------------------------------
  // RECORD
  // -------------------------------------------------------------
  private encodeRecord(node: Node, indent: number): string {
    const innerMeta = this.metaWrapped(node);

    const parts: string[] = [];
    if (node.schema.fields && node.schema.fields.length > 0) {
      for (const fieldDef of node.schema.fields) {
        const fieldNode = node.fields[fieldDef.name];
        if (fieldNode) {
          let val = this.encode(fieldNode, indent - this.config.startIndent, false).trim();
          val = this.applyTypeTag(val, fieldNode.schema, fieldDef);
          parts.push(val);
        } else {
          parts.push(this.c('null', Colors.NULL));
        }
      }
      const sep = this.config.compact ? ',' : ', ';
      return '(' + innerMeta + parts.join(sep) + ')';
    } else {
      return '(' + innerMeta + this.c('null', Colors.NULL) + ')';
    }
  }
}
